gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java

```diff
@@ -0,0 +1,179 @@
+      @Override
+      public Limiter call() throws Exception {
+        return new RateBasedLimiter(limitPerSecond);
+      }
+    });
+  }
+
+  @Override
+  public boolean delete(Path path) throws IOException {
+    this.acquirePermit();
+    return this.fs.delete(path);
+  }
+
+  @Override
+  public boolean delete(Path path, boolean recursive) throws IOException {
+    this.acquirePermit();
+    return this.fs.delete(path, recursive);
+  }
+
+  @Override
+  public boolean exists(Path path) throws IOException {
+    this.acquirePermit();
+    return this.fs.equals(path);
+  }
+
+  @Override
+  public FileStatus getFileStatus(Path path) throws IOException {
+    this.acquirePermit();
+    return this.fs.getFileStatus(path);
+  }
+
+  @Override
+  public FileStatus[] globStatus(Path pathPattern) throws IOException {
+    this.acquirePermit();
+    return this.fs.globStatus(pathPattern);
+  }
+
+  @Override
+  public FileStatus[] listStatus(Path path) throws IOException {
+    this.acquirePermit();
+    return this.fs.listStatus(path);
+  }
+
+  @Override
+  public FileStatus[] listStatus(Path path, PathFilter filter) throws IOException {
+    this.acquirePermit();
+    return this.fs.listStatus(path, filter);
+  }
+
+  @Override
+  public boolean mkdirs(Path path, FsPermission permission) throws IOException {
+    this.acquirePermit();
+    return this.fs.mkdirs(path, permission);
+  }
+
+  @Override
+  public boolean mkdirs(Path path) throws IOException {
+    this.acquirePermit();
+    return this.fs.mkdirs(path);
+  }
+
+  @Override
+  public boolean rename(Path path0, Path path1) throws IOException {
+    this.acquirePermit();
+    return this.fs.rename(path0, path1);
+  }
+
+  @Override
+  public FSDataOutputStream append(Path path, int bufferSize, Progressable progress) throws IOException {
+    this.acquirePermit();
+    return this.fs.append(path, bufferSize, progress);
+  }
+
+  @Override
+  public FSDataOutputStream create(Path path, FsPermission permission, boolean overwrite, int bufferSize, short replication, long blockSize,
+      Progressable progress) throws IOException {
+    this.acquirePermit();
+    return this.fs.create(path, permission, overwrite, bufferSize, replication, blockSize, progress);
+  }
+
+  @Override
+  public URI getUri() {
+    return this.fs.getUri();
+  }
+
+  @Override
+  public Path getWorkingDirectory() {
+    return this.fs.getWorkingDirectory();
+  }
+
+  @Override
+  public FSDataInputStream open(Path path, int bufferSize) throws IOException {
+    this.acquirePermit();
+    return this.fs.open(path, bufferSize);
+  }
+
+  @Override
+  public void setWorkingDirectory(Path path) {
+    this.fs.setWorkingDirectory(path);
+  }
+
+  public void startRateControl() {
+    this.rateLimiter.start();
+  }
+
+  public void stopRateControl() {
+    this.rateLimiter.stop();
+  }
+
+  private void acquirePermit() throws IOException {
```

Comment:
I would rather not store the rateLimiter as an instance variable, instead getting it from the cache every time acquirePermit() is called. This way, if the rateLimiter ever gets cleaned from cache, we will not end up with a new file system in the same uri using a different rate limiter.